File: Yow.CoD.Finance.Domain.Tests/Specification.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using NUnit.Framework;
using Yow.CoD.Finance.Domain.Contracts;
using Yow.CoD.Finance.Domain.Model;
using Yow.CoD.Finance.Domain.Services;

namespace Yow.CoD.Finance.Domain.Tests
{
    [TestFixture]
    public abstract class Specification<T, TCommand, TReceipt> 
        where T : AggregateRoot 
        where TCommand : Command
        where TReceipt : Receipt
    {
        private FakeRepository<T> _repository;
        protected abstract IEnumerable<Event> Given();
        protected abstract TCommand When();
        protected abstract IHandler<TCommand, TReceipt> CreateHandler();

        protected T Sut { get; set; }
        protected List<Event> Produced { get; private set; }
        protected TReceipt Receipt { get; private set; }
        protected Exception Caught { get; private set; }

        protected IRepository<T> Repository => _repository;

        [SetUp]
        public async Task SetUp()
        {
            try
            {
                _repository = new FakeRepository<T>();
                Sut = await Repository.Get(Guid.NewGuid());
                var initialState = Given().ToArray();
                
                foreach (var @event in initialState)
                {
                    Sut.ApplyEvent(@event);
                }
                
                var handler = CreateHandler();
                var cmd = When();
                Receipt = await handler.Handle(cmd);
                Produced = _repository.CommitedEvents;
            }
            catch (Exception e)
            {
                Caught = e;
            }
        }
    }
}

File: typescript/packages/domain-tests/src/specification.ts
import {
  AggregateRoot,
  type Command,
  Event,
  type Handler,
  type Receipt,
} from "@cod/domain";
import { FakeRepository } from "./fake-repository.js";

export abstract class Specification<
  TAggregate extends AggregateRoot,
  TCommand extends Command,
  TReceipt extends Receipt,
> {
  #repository!: FakeRepository<TAggregate>;
  #sut!: TAggregate;
  #produced: Event[] = [];
  #receipt?: TReceipt;
  #caught?: Error;

  protected abstract createAggregate(id: string): TAggregate;
  protected abstract given(): Event[];
  protected abstract when(): TCommand;
  protected abstract createHandler(): Handler<TCommand, TReceipt>;

  get sut(): TAggregate {
    return this.#sut;
  }

  get produced(): Event[] {
    return this.#produced;
  }

  get receipt(): TReceipt | undefined {
    return this.#receipt;
  }

  get caught(): Error | undefined {
    return this.#caught;
  }

  get repository(): FakeRepository<TAggregate> {
    return this.#repository;
  }

  async execute(): Promise<void> {
    try {
      this.#repository = new FakeRepository<TAggregate>((id) =>
        this.createAggregate(id),
      );
      this.#sut = await this.#repository.get(crypto.randomUUID());

      for (const event of this.given()) {
        this.#sut.applyEvent(event);
      }

      const handler = this.createHandler();
      const command = this.when();
      this.#receipt = await handler.handle(command);
      this.#produced = this.#repository.committedEvents;
    } catch (e) {
      this.#caught = e as Error;
    }
  }
}
